jwt/validate.go

```diff
@@ -44,145 +44,282 @@ func timeClaim(t Token, clock Clock, c string) time.Time {
 // See the various `WithXXX` functions for optional parameters
 // that can control the behavior of this method.
 func Validate(t Token, options ...ValidateOption) error {
-	var issuer string
-	var subject string
-	var audience string
-	var jwtid string
 	var clock Clock = ClockFunc(time.Now)
 	var skew time.Duration
-	var deltas []delta
-	requiredMap := make(map[string]struct{})
-	claimValues := make(map[string]interface{})
+	var validators = []Validator{
+		IsIssuedAtValid(),
+		IsExpirationValid(),
+		IsNbfValid(),
+	}
 	for _, o := range options {
 		//nolint:forcetypeassert
 		switch o.Ident() {
 		case identClock{}:
 			clock = o.Value().(Clock)
 		case identAcceptableSkew{}:
 			skew = o.Value().(time.Duration)
-		case identIssuer{}:
-			issuer = o.Value().(string)
-		case identSubject{}:
-			subject = o.Value().(string)
-		case identAudience{}:
-			audience = o.Value().(string)
-		case identJwtid{}:
-			jwtid = o.Value().(string)
-		case identRequiredClaim{}:
-			requiredMap[o.Value().(string)] = struct{}{}
-		case identTimeDelta{}:
-			d := o.Value().(delta)
-			deltas = append(deltas, d)
-			if d.c1 != "" {
-				if err := isSupportedTimeClaim(d.c1); err != nil {
-					return err
+		case identValidator{}:
+			v := o.Value().(Validator)
+			switch v := v.(type) {
+			case *isInTimeRange:
+				if v.c1 != "" {
+					if err := isSupportedTimeClaim(v.c1); err != nil {
+						return err
+					}
+					validators = append(validators, IsRequired(v.c1))
 				}
-				requiredMap[d.c1] = struct{}{}
-			}
-
-			if d.c2 != "" {
-				if err := isSupportedTimeClaim(d.c2); err != nil {
-					return err
+				if v.c2 != "" {
+					if err := isSupportedTimeClaim(v.c2); err != nil {
+						return err
+					}
+					validators = append(validators, IsRequired(v.c2))
 				}
-				requiredMap[d.c2] = struct{}{}
 			}
-		case identClaim{}:
-			claim := o.Value().(claimValue)
-			claimValues[claim.name] = claim.value
+			validators = append(validators, v)
 		}
 	}
 
-	for c := range requiredMap {
-		if _, ok := t.Get(c); !ok {
-			return errors.Errorf(`required claim %s was not found`, c)
+	ctx := SetValidationCtxSkew(context.Background(), skew)
```

Comment:
Is it possible to propagate the parent context?

Reply:
hmmm, no, because `jwt.Validate()` does not take a context :/

Reply:
Is there any new API?

```go
jwt.Validate(
  jwt.ValidateWithContext(ctx),
)
```

Or make jwt.Validate depreacated
```go
jwt.ValidateWithContext(ctx, ...)
```

Reply:
Sure, it's _doable_ but why? is there a usecase?

Reply:
I considered some usecases for claim validation logic.
- RPC call to another server over network
- Add span with tracing framework such as OpenTelemetry

Reply:
Ah, the span, I can sympathize with...

Reply:
fixed in bb7f584